libraries/abstractions/pkcs11/test/iot_test_pkcs11.c

```diff
@@ -700,7 +734,56 @@ TEST( Full_PKCS11_StartFinish, AFQP_OpenSessionCloseSession )
     TEST_ASSERT_EQUAL_MESSAGE( CKR_CRYPTOKI_NOT_INITIALIZED, xResult, "Negative Test: Opened a session before initializing module." );
 }
 
+/*--------------------------------------------------------*/
+/*-------------- Capabilities Tests --------------------- */
+/*--------------------------------------------------------*/
+
+TEST( Full_PKCS11_Capabilities, AFQP_Capabilities )
+{
+    CK_RV xResult = 0;
+    CK_ULONG xSlotCount = 0;
+    CK_SLOT_ID_PTR pxSlotId = NULL;
+    CK_MECHANISM_INFO MechanismInfo = { 0 };
+
+    /* Determine the number of slots. */
+    xResult = pxGlobalFunctionList->C_GetSlotList( CK_TRUE, NULL, &xSlotCount );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to get slot count" );
+
+    /* Allocate memory to receive the list of slots, plus one extra. */
+    pxSlotId = pvPortMalloc( sizeof( CK_SLOT_ID ) * xSlotCount );
+    TEST_ASSERT_NOT_EQUAL_MESSAGE( NULL, pxSlotId, "Failed malloc memory for slot list" );
+
+    /* Call C_GetSlotList again to receive all slots with tokens present. */
+    xResult = pxGlobalFunctionList->C_GetSlotList( CK_TRUE, pxSlotId, &xSlotCount );
+    TEST_ASSERT_EQUAL_MESSAGE( CKR_OK, xResult, "Failed to get slot count" );
+
+    /* Check for RSA PKCS #1 signing support. */
+    xResult = pxGlobalFunctionList->C_GetMechanismInfo( pxSlotId[ 0 ], CKM_RSA_PKCS, &MechanismInfo );
+    TEST_ASSERT_TRUE( CKR_OK == xResult || CKR_MECHANISM_INVALID == xResult );
+
+    if( CKR_OK == xResult )
+    {
+        configPRINTF( ( "The PKCS #11 module supports RSA signing.\r\n" ) );
+    }
+
+    /* Check for ECDSA support. */
+    xResult = pxGlobalFunctionList->C_GetMechanismInfo( pxSlotId[ 0 ], CKM_ECDSA, &MechanismInfo );
+    TEST_ASSERT_TRUE( CKR_OK == xResult || CKR_MECHANISM_INVALID == xResult );
 
+    if( CKR_OK == xResult )
+    {
+        configPRINTF( ( "The PKCS #11 module supports ECDSA.\r\n" ) );
+    }
+
+    /* Check for elliptic-curve key generation support. */
+    xResult = pxGlobalFunctionList->C_GetMechanismInfo( pxSlotId[ 0 ], CKM_EC_KEY_PAIR_GEN, &MechanismInfo );
+    TEST_ASSERT_TRUE( CKR_OK == xResult || CKR_MECHANISM_INVALID == xResult );
+
+    if( CKR_OK == xResult )
+    {
+        configPRINTF( ( "The PKCS #11 module supports elliptic-curve key generation.\r\n" ) );
+    }
```

Comment:
Should we also assert that P256 and 2048 rsa are within the key size info ranges?